File: training-data/rust/huffman_coding.rs
use std::collections::{BinaryHeap, HashMap};
use std::cmp::Ordering;

#[derive(Eq, PartialEq)]
struct HuffNode {
    ch: Option<char>,
    freq: i32,
    left: Option<Box<HuffNode>>,
    right: Option<Box<HuffNode>>,
}

impl Ord for HuffNode {
    fn cmp(&self, other: &Self) -> Ordering {
        other.freq.cmp(&self.freq)
    }
}

impl PartialOrd for HuffNode {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn build_tree(text: &str) -> HuffNode {
    let mut counts = HashMap::new();
    for ch in text.chars() {
        *counts.entry(ch).or_insert(0) += 1;
    }
    let mut heap: BinaryHeap<HuffNode> = counts
        .into_iter()
        .map(|(ch, freq)| HuffNode { ch: Some(ch), freq, left: None, right: None })
        .collect();

    while heap.len() > 1 {
        let a = heap.pop().unwrap();
        let b = heap.pop().unwrap();
        heap.push(HuffNode { ch: None, freq: a.freq + b.freq, left: Some(Box::new(a)), right: Some(Box::new(b)) });
    }
    heap.pop().unwrap()
}

fn build_codes(node: &HuffNode, prefix: String, codes: &mut HashMap<char, String>) {
    if let Some(ch) = node.ch {
        codes.insert(ch, if prefix.is_empty() { "0".to_string() } else { prefix });
        return;
    }
    if let Some(left) = &node.left {
        build_codes(left, format!("{}0", prefix), codes);
    }
    if let Some(right) = &node.right {
        build_codes(right, format!("{}1", prefix), codes);
    }
}

fn encode(text: &str) -> String {
    let tree = build_tree(text);
    let mut codes = HashMap::new();
    build_codes(&tree, String::new(), &mut codes);
    text.chars().map(|ch| codes[&ch].clone()).collect()
}

fn main() {
    println!("{}", encode("abracadabra"));
}

File: training-data/scala/HuffmanCoding.scala
import scala.collection.mutable

sealed trait HuffmanTree { def freq: Int }
case class Leaf(char: Char, freq: Int) extends HuffmanTree
case class Node(left: HuffmanTree, right: HuffmanTree, freq: Int) extends HuffmanTree

object HuffmanCoding {
  def buildTree(freqs: Map[Char, Int]): HuffmanTree = {
    val pq = mutable.PriorityQueue[HuffmanTree]()(Ordering.by((t: HuffmanTree) => t.freq).reverse)
    freqs.foreach { case (ch, f) => pq.enqueue(Leaf(ch, f)) }

    while (pq.size > 1) {
      val a = pq.dequeue()
      val b = pq.dequeue()
      pq.enqueue(Node(a, b, a.freq + b.freq))
    }
    pq.dequeue()
  }

  def codes(tree: HuffmanTree, prefix: String = ""): Map[Char, String] = tree match {
    case Leaf(ch, _) => Map(ch -> (if (prefix.isEmpty) "0" else prefix))
    case Node(left, right, _) => codes(left, prefix + "0") ++ codes(right, prefix + "1")
  }

  def main(args: Array[String]): Unit = {
    val text = "abracadabra"
    val freqs = text.groupBy(identity).view.mapValues(_.length).toMap
    val tree = buildTree(freqs)
    val table = codes(tree)
    table.toSeq.sortBy(_._1).foreach { case (ch, code) => println(s"$ch -> $code") }
  }
}
